\documentclass[]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
  \newcommand{\euro}{â‚¬}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}


\usepackage{longtable,booktabs}
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

\RequirePackage[]{/Library/Frameworks/R.framework/Versions/3.4/Resources/library/BiocStyle/resources/tex/Bioconductor2}

% Create subtitle command for use in maketitle
\newcommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}
\bioctitle[]{Loci2path: regulatory annotation of genomic intervals based on
tissue-specific expression QTLs}
  \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
\author{Tianlei Xu and Zhaohui Qin}
  \preauthor{\centering\large\emph}
  \postauthor{\par}
  \predate{\centering\large\emph}
  \postdate{\par}
  \date{19 August 2017}


% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

% code highlighting
\definecolor{fgcolor}{rgb}{0.251, 0.251, 0.251}
\newcommand{\hlnum}[1]{\textcolor[rgb]{0.816,0.125,0.439}{#1}}%
\newcommand{\hlstr}[1]{\textcolor[rgb]{0.251,0.627,0.251}{#1}}%
\newcommand{\hlcom}[1]{\textcolor[rgb]{0.502,0.502,0.502}{\textit{#1}}}%
\newcommand{\hlopt}[1]{\textcolor[rgb]{0,0,0}{#1}}%
\newcommand{\hlstd}[1]{\textcolor[rgb]{0.251,0.251,0.251}{#1}}%
\newcommand{\hlkwa}[1]{\textcolor[rgb]{0.125,0.125,0.941}{#1}}%
\newcommand{\hlkwb}[1]{\textcolor[rgb]{0,0,0}{#1}}%
\newcommand{\hlkwc}[1]{\textcolor[rgb]{0.251,0.251,0.251}{#1}}%
\newcommand{\hlkwd}[1]{\textcolor[rgb]{0.878,0.439,0.125}{#1}}%
\let\hlipl\hlkwb
%
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
%
\newenvironment{Shaded}{\begin{myshaded}}{\end{myshaded}}
% set background for result chunks
\let\oldverbatim\verbatim
\renewenvironment{verbatim}{\color{codecolor}\begin{myshaded}\begin{oldverbatim}}{\end{oldverbatim}\end{myshaded}}
%
\newcommand{\KeywordTok}[1]{\hlkwd{#1}}
\newcommand{\DataTypeTok}[1]{\hlkwc{#1}}
\newcommand{\DecValTok}[1]{\hlnum{#1}}
\newcommand{\BaseNTok}[1]{\hlnum{#1}}
\newcommand{\FloatTok}[1]{\hlnum{#1}}
\newcommand{\CharTok}[1]{\hlstr{#1}}
\newcommand{\StringTok}[1]{\hlstr{#1}}
\newcommand{\CommentTok}[1]{\hlcom{#1}}
\newcommand{\OtherTok}[1]{{#1}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{{#1}}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{{#1}}}
\newcommand{\RegionMarkerTok}[1]{{#1}}
\newcommand{\ErrorTok}[1]{\textbf{{#1}}}
\newcommand{\NormalTok}[1]{\hlstd{#1}}
%
\AtBeginDocument{\bibliographystyle{/Library/Frameworks/R.framework/Versions/3.4/Resources/library/BiocStyle/resources/tex/unsrturl}}

\usepackage{amsthm}
\newtheorem{theorem}{Theorem}[section]
\newtheorem{lemma}{Lemma}[section]
\theoremstyle{definition}
\newtheorem{definition}{Definition}[section]
\newtheorem{corollary}{Corollary}[section]
\newtheorem{proposition}{Proposition}[section]
\theoremstyle{definition}
\newtheorem{example}{Example}[section]
\theoremstyle{remark}
\newtheorem*{remark}{Remark}
\begin{document}
\maketitle
\begin{abstract}
Annotating a given genomic locus or a set of genomic loci is an
important yet challenging task. This is especially true for the
non-coding part of the genome which is enormous yet poorly understood.
Since gene set enrichment analyses have demonstrated to be effective
approach to annotate a set of genes, this idea can be extended to
explore the enrichment of functional elements or features in a set of
genomic intervals to reveal potential functional connections. In this
study, we describe a novel computational strategy that takes advantage
of the newly emerged, genome-wide and tissue-specific expression
quantitative trait loci (eQTL) information to help annotate a set of
genomic intervals in terms of transcription regulation. By checking the
presence or absence of millions of eQTLs in the set of genomic intervals
of interest, loci2path build a bridge connecting genomic intervals to
biological pathway or pre-defined biological-meaningful gene sets. Our
method enjoys two key advantages over existing methods: first, we no
longer rely on proximity to link a locus to a gene which has shown to be
unreliable; second, eQTL allows us to provide the regulatory annotation
under the context of specific tissue types which is important.
\end{abstract}

\packageVersion{loci2path 0.1.0}

{
\setcounter{tocdepth}{2}
\tableofcontents
\newpage
}
\section{Prepare input dataset for
query}\label{prepare-input-dataset-for-query}

\subsection{Query regions}\label{query-regions}

\texttt{loci2path} takes query regions in the format of
\texttt{GenomicRanges}. Only the Genomic Locations (chromosomes, start
and end position) will be used. Strand information and other metadata
columns are ignored. In the demo data, 47 regions associated with
Psoriasis disease were downloaded from \textbf{immunoBase.org} and used
as demo query regions.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{bed.file=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"query/Psoriasis.BED"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"loci2path"}\NormalTok{)}
\NormalTok{query.bed=}\KeywordTok{read.table}\NormalTok{(bed.file, }\DataTypeTok{header=}\NormalTok{F)}
\KeywordTok{colnames}\NormalTok{(query.bed)=}\KeywordTok{c}\NormalTok{(}\StringTok{"chr"}\NormalTok{,}\StringTok{"start"}\NormalTok{,}\StringTok{"end"}\NormalTok{)}
\NormalTok{query.gr=}\KeywordTok{makeGRangesFromDataFrame}\NormalTok{(query.bed)}
\end{Highlighting}
\end{Shaded}

\subsection{Prepare eQTL sets.}\label{prepare-eqtl-sets.}

eQTL sets are entities recording 1-to-1 links between eQTL SNPs and
genes. eQTL set entity also contains the following information: tissue
name for the eQTL study, IDs and genomic ranges for the eQTL SNPs, IDs
for the associated genes.

eQTL set can be constructed manually by specifying the corresponding
information in each slot.

eQTL set list is a list of multiple eQTL sets, usually collected from
different tissues.

Below is an example to construct customized eQTL set and eQTL set list
using demo data files. In the demo data folder, three eQTL sets
downloaded from GTEx project are included. Due to the large size, each
eQTL dataset is down sampled to 3000 records for demostration purpose.

\subsubsection{construct eQTL set}\label{construct-eqtl-set}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{brain.file=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"eqtl/brain.gtex.txt"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"loci2path"}\NormalTok{)}
\NormalTok{tab=}\KeywordTok{read.table}\NormalTok{(brain.file, }\DataTypeTok{stringsAsFactors =} \NormalTok{F, }\DataTypeTok{header =} \NormalTok{T)}
\NormalTok{snp.gr=}\KeywordTok{GRanges}\NormalTok{(}\DataTypeTok{seqnames=}\KeywordTok{Rle}\NormalTok{(tab$snp.chr), }
  \DataTypeTok{ranges=}\KeywordTok{IRanges}\NormalTok{(}\DataTypeTok{start=}\NormalTok{tab$snp.pos, }
  \DataTypeTok{width=}\DecValTok{1}\NormalTok{))}
\NormalTok{brain.eset=}\KeywordTok{eqtlSet}\NormalTok{(}\DataTypeTok{tissue=}\StringTok{"brain"}\NormalTok{,}
  \DataTypeTok{snp.id=}\NormalTok{tab$snp.id,}
  \DataTypeTok{snp.gr=}\NormalTok{snp.gr,}
  \DataTypeTok{gene=}\KeywordTok{as.character}\NormalTok{(tab$gene.entrez.id))}
\NormalTok{brain.eset}
\NormalTok{## An object of class eqtlSet}
\NormalTok{##  eQTL collected from tissue: brain }
\NormalTok{##  number of eQTLs: 3000 }
\NormalTok{##  number of associated genes: 815}

\NormalTok{skin.file=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"eqtl/skin.gtex.txt"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"loci2path"}\NormalTok{)}
\NormalTok{tab=}\KeywordTok{read.table}\NormalTok{(skin.file, }\DataTypeTok{stringsAsFactors =} \NormalTok{F, }\DataTypeTok{header =} \NormalTok{T)}
\NormalTok{snp.gr=}\KeywordTok{GRanges}\NormalTok{(}\DataTypeTok{seqnames=}\KeywordTok{Rle}\NormalTok{(tab$snp.chr), }
  \DataTypeTok{ranges=}\KeywordTok{IRanges}\NormalTok{(}\DataTypeTok{start=}\NormalTok{tab$snp.pos, }
  \DataTypeTok{width=}\DecValTok{1}\NormalTok{))}
\NormalTok{skin.eset=}\KeywordTok{eqtlSet}\NormalTok{(}\DataTypeTok{tissue=}\StringTok{"skin"}\NormalTok{,}
  \DataTypeTok{snp.id=}\NormalTok{tab$snp.id,}
  \DataTypeTok{snp.gr=}\NormalTok{snp.gr,}
  \DataTypeTok{gene=}\KeywordTok{as.character}\NormalTok{(tab$gene.entrez.id))}
\NormalTok{skin.eset}
\NormalTok{## An object of class eqtlSet}
\NormalTok{##  eQTL collected from tissue: skin }
\NormalTok{##  number of eQTLs: 3000 }
\NormalTok{##  number of associated genes: 1588}

\NormalTok{blood.file=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"eqtl/blood.gtex.txt"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"loci2path"}\NormalTok{)}
\NormalTok{tab=}\KeywordTok{read.table}\NormalTok{(blood.file, }\DataTypeTok{stringsAsFactors =} \NormalTok{F, }\DataTypeTok{header =} \NormalTok{T)}
\NormalTok{snp.gr=}\KeywordTok{GRanges}\NormalTok{(}\DataTypeTok{seqnames=}\KeywordTok{Rle}\NormalTok{(tab$snp.chr), }
  \DataTypeTok{ranges=}\KeywordTok{IRanges}\NormalTok{(}\DataTypeTok{start=}\NormalTok{tab$snp.pos, }
  \DataTypeTok{width=}\DecValTok{1}\NormalTok{))}
\NormalTok{blood.eset=}\KeywordTok{eqtlSet}\NormalTok{(}\DataTypeTok{tissue=}\StringTok{"blood"}\NormalTok{,}
  \DataTypeTok{snp.id=}\NormalTok{tab$snp.id,}
  \DataTypeTok{snp.gr=}\NormalTok{snp.gr,}
  \DataTypeTok{gene=}\KeywordTok{as.character}\NormalTok{(tab$gene.entrez.id))}
\NormalTok{blood.eset}
\NormalTok{## An object of class eqtlSet}
\NormalTok{##  eQTL collected from tissue: blood }
\NormalTok{##  number of eQTLs: 3000 }
\NormalTok{##  number of associated genes: 1419}
\end{Highlighting}
\end{Shaded}

\subsubsection{construct eQTL set list}\label{construct-eqtl-set-list}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{eset.list=}\KeywordTok{list}\NormalTok{(}\DataTypeTok{Brain=}\NormalTok{brain.eset, }\DataTypeTok{Skin=}\NormalTok{skin.eset, }\DataTypeTok{Blood=}\NormalTok{blood.eset)}
\NormalTok{eset.list}
\NormalTok{## $Brain}
\NormalTok{## An object of class eqtlSet}
\NormalTok{##  eQTL collected from tissue: brain }
\NormalTok{##  number of eQTLs: 3000 }
\NormalTok{##  number of associated genes: 815 }
\NormalTok{## }
\NormalTok{## $Skin}
\NormalTok{## An object of class eqtlSet}
\NormalTok{##  eQTL collected from tissue: skin }
\NormalTok{##  number of eQTLs: 3000 }
\NormalTok{##  number of associated genes: 1588 }
\NormalTok{## }
\NormalTok{## $Blood}
\NormalTok{## An object of class eqtlSet}
\NormalTok{##  eQTL collected from tissue: blood }
\NormalTok{##  number of eQTLs: 3000 }
\NormalTok{##  number of associated genes: 1419}
\end{Highlighting}
\end{Shaded}

\subsection{Prepare gene set
collection}\label{prepare-gene-set-collection}

A geneset collection contains a list of gene sets, with each gene set is
represented as a vector of member genes. A vector of description is also
provided as the metadata slot for each gene set. The total number of
gene in the geneset collection is also required to perform the
enrichment test. In this tutorial the BIOCARTA pathway collection was
downloaded from MSigDB.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{biocarta.link.file=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"geneSet/biocarta.txt"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"loci2path"}\NormalTok{)}
\NormalTok{biocarta.set.file=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"geneSet/biocarta.set.txt"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"loci2path"}\NormalTok{)}

\NormalTok{biocarta.link=}\KeywordTok{read.delim}\NormalTok{(biocarta.link.file, }\DataTypeTok{header =} \NormalTok{F, }\DataTypeTok{stringsAsFactors =} \NormalTok{F)}
\NormalTok{set.geneid=}\KeywordTok{read.table}\NormalTok{(biocarta.set.file, }\DataTypeTok{stringsAsFactors =} \NormalTok{F)}
\NormalTok{set.geneid=}\KeywordTok{strsplit}\NormalTok{(set.geneid[,}\DecValTok{1}\NormalTok{], }\DataTypeTok{split=}\StringTok{","}\NormalTok{)}
\KeywordTok{names}\NormalTok{(set.geneid)=biocarta.link[,}\DecValTok{1}\NormalTok{]}

\KeywordTok{head}\NormalTok{(biocarta.link)}
\NormalTok{##                              V1}
\NormalTok{## 1         BIOCARTA_RELA_PATHWAY}
\NormalTok{## 2          BIOCARTA_NO1_PATHWAY}
\NormalTok{## 3          BIOCARTA_CSK_PATHWAY}
\NormalTok{## 4      BIOCARTA_SRCRPTP_PATHWAY}
\NormalTok{## 5          BIOCARTA_AMI_PATHWAY}
\NormalTok{## 6 BIOCARTA_GRANULOCYTES_PATHWAY}
\NormalTok{##                                                                              V2}
\NormalTok{## 1         http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_RELA_PATHWAY}
\NormalTok{## 2          http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_NO1_PATHWAY}
\NormalTok{## 3          http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_CSK_PATHWAY}
\NormalTok{## 4      http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_SRCRPTP_PATHWAY}
\NormalTok{## 5          http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_AMI_PATHWAY}
\NormalTok{## 6 http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_GRANULOCYTES_PATHWAY}
\KeywordTok{head}\NormalTok{(set.geneid)}
\NormalTok{## $BIOCARTA_RELA_PATHWAY}
\NormalTok{##  [1] "8517" "1147" "2033" "5970" "7124" "3551" "7133" "8841" "7132" "7189"}
\NormalTok{## [11] "8772" "1387" "8737" "4790" "4792" "8717"}
\NormalTok{## }
\NormalTok{## $BIOCARTA_NO1_PATHWAY}
\NormalTok{##  [1] "5140"   "805"    "58"     "124827" "801"    "5577"   "3827"   "6262"  }
\NormalTok{##  [9] "1128"   "7422"   "3320"   "6541"   "5139"   "5138"   "624"    "147908"}
\NormalTok{## [17] "121916" "4846"   "1134"   "2321"   "3791"   "5567"   "7135"   "5568"  }
\NormalTok{## [25] "2324"   "857"    "207"    "5573"   "5576"   "5575"   "808"    "5592"  }
\NormalTok{## [33] "5593"  }
\NormalTok{## }
\NormalTok{## $BIOCARTA_CSK_PATHWAY}
\NormalTok{##  [1] "7535" "1445" "920"  "5577" "5567" "915"  "5568" "916"  "917"  "2778"}
\NormalTok{## [11] "2792" "6957" "2782" "6955" "5573" "5576" "919"  "1387" "5575" "107" }
\NormalTok{## [21] "3932" "5788" "3123" "3122"}
\NormalTok{## }
\NormalTok{## $BIOCARTA_SRCRPTP_PATHWAY}
\NormalTok{##  [1] "1445" "6714" "994"  "995"  "5579" "2885" "993"  "5578" "891"  "5786"}
\NormalTok{## [11] "983" }
\NormalTok{## }
\NormalTok{## $BIOCARTA_AMI_PATHWAY}
\NormalTok{##  [1] "2159"  "7035"  "2147"  "1282"  "2149"  "1284"  "1285"  "1286"  "5627" }
\NormalTok{## [10] "2266"  "5624"  "2243"  "5340"  "462"   "2244"  "5327"  "1288"  "51327"}
\NormalTok{## [19] "1287"  "2155" }
\NormalTok{## }
\NormalTok{## $BIOCARTA_GRANULOCYTES_PATHWAY}
\NormalTok{##  [1] "5175" "7124" "3552" "3683" "3684" "3383" "6402" "6403" "3458" "6404"}
\NormalTok{## [11] "727"  "3689" "1440" "3576"}
\end{Highlighting}
\end{Shaded}

In order to build gene set, we also need to know the total number of
genes in order to perform enrichment test. In this study, the total
number of gene in MSigDB pathway collection is 31,847 (ref)

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#build geneSet}
\NormalTok{biocarta=}\KeywordTok{geneSet}\NormalTok{(}
  \DataTypeTok{gene.set=}\NormalTok{set.geneid,}
  \DataTypeTok{description=}\NormalTok{biocarta.link[,}\DecValTok{2}\NormalTok{],}
  \DataTypeTok{total.number.gene=}\DecValTok{31847}\NormalTok{)}
\NormalTok{biocarta}
\NormalTok{## An object of class geneSet}
\NormalTok{##  Number of gene sets: 217 }
\NormalTok{##     6 ~ 87  genes within sets}
\end{Highlighting}
\end{Shaded}

\section{Perform query}\label{perform-query}

\section{explore query result}\label{explore-query-result}

\subsection{extract tissue-pathway
heatmap}\label{extract-tissue-pathway-heatmap}

\subsection{extract word cloud from
result}\label{extract-word-cloud-from-result}

\subsection{obtain eQTL gene list}\label{obtain-eqtl-gene-list}

\subsection{obtain average tissue degree for each
pathway}\label{obtain-average-tissue-degree-for-each-pathway}

\subsection{obtain tissue enrichment for query
regions}\label{obtain-tissue-enrichment-for-query-regions}

\begin{verbatim}
## GRanges object with 47 ranges and 0 metadata columns:
##        seqnames                 ranges strand
##           <Rle>              <IRanges>  <Rle>
##    [1]     chr1 [  8200690,   8306031]      *
##    [2]     chr1 [152536784, 152785813]      *
##    [3]     chr1 [ 24461438,  24527816]      *
##    [4]     chr1 [ 67594559,  67767993]      *
##    [5]     chr1 [ 25224957,  25308276]      *
##    ...      ...                    ...    ...
##   [43]    chr18   [52210075, 52409477]      *
##   [44]    chr19   [10634264, 11164781]      *
##   [45]    chr19   [10390709, 10628548]      *
##   [46]    chr20   [48408615, 48662582]      *
##   [47]    chr22   [21809185, 22003928]      *
##   -------
##   seqinfo: 16 sequences from an unspecified genome; no seqlengths
## $Brain_Cortex
## An object of class eqtlSet
##  eQTL collected from tissue: Brain_Cortex 
##  number of eQTLs: 131424 
##  number of associated genes: 1796 
## 
## $Skin_Sun_Exposed_Lower_leg
## An object of class eqtlSet
##  eQTL collected from tissue: Skin_Sun_Exposed_Lower_leg 
##  number of eQTLs: 712745 
##  number of associated genes: 6171 
## 
## $Whole_Blood
## An object of class eqtlSet
##  eQTL collected from tissue: Whole_Blood 
##  number of eQTLs: 594632 
##  number of associated genes: 5073
## An object of class geneSet
##  Number of gene sets: 217 
##     6 ~ 87  genes within sets
## Start query: 3 eqtl Sets...
## 1 of 3: Brain_Cortex...
## 2 of 3: Skin_Sun_Exposed_Lower_leg...
## 3 of 3: Whole_Blood...
## 
## done!
##                       tissue                     name_pthw eQTL_pthw
## 1                Whole_Blood     BIOCARTA_MONOCYTE_PATHWAY       157
## 2                Whole_Blood       BIOCARTA_LECTIN_PATHWAY      3227
## 3                Whole_Blood BIOCARTA_GRANULOCYTES_PATHWAY        96
## 4                Whole_Blood      BIOCARTA_CLASSIC_PATHWAY      3233
## 5 Skin_Sun_Exposed_Lower_leg       BIOCARTA_LECTIN_PATHWAY      2859
## 6                Whole_Blood         BIOCARTA_COMP_PATHWAY      3325
##   eQTL_total_tissue eQTL_query eQTL_pthw_query log_ratio pval_lr
## 1            594632      11943              14  1.490609      NA
## 2            594632      11943             538  2.116348      NA
## 3            594632      11943              14  1.982507      NA
## 4            594632      11943             538  2.114490      NA
## 5            712745      17062             262  1.342387      NA
## 6            594632      11943             538  2.086431      NA
##     pval_fisher num_gene_set num_gene_query num_gene_hit  gene_hit
## 1  4.108902e-06           11             77            2 3684;3383
## 2 8.096423e-312           12             77            2   721;720
## 3  8.812780e-09           14             77            2 3684;3383
## 4 2.149518e-311           14             77            2   721;720
## 5  2.372477e-74           12            108            2   721;720
## 6 5.197609e-305           19             77            2   721;720
##   log_ratio_gene pval_fisher_gene
## 1       4.320145     0.0003129003
## 2       4.233134     0.0003748918
## 3       4.078983     0.0005152773
## 4       4.078983     0.0005152773
## 5       3.894808     0.0007355142
## 6       3.773601     0.0009607091
## Start query: 3 eqtl Sets...
## Run in parallel mode...
## 
## done!
##                       name_pthw eQTL_pthw eQTL_total_tissue eQTL_query
## 1     BIOCARTA_MONOCYTE_PATHWAY       157            594632      11943
## 2       BIOCARTA_LECTIN_PATHWAY      3227            594632      11943
## 3 BIOCARTA_GRANULOCYTES_PATHWAY        96            594632      11943
## 4      BIOCARTA_CLASSIC_PATHWAY      3233            594632      11943
## 5       BIOCARTA_LECTIN_PATHWAY      2859            712745      17062
## 6         BIOCARTA_COMP_PATHWAY      3325            594632      11943
##   eQTL_pthw_query log_ratio pval_lr   pval_fisher num_gene_set
## 1              14  1.490609      NA  4.108902e-06           11
## 2             538  2.116348      NA 8.096423e-312           12
## 3              14  1.982507      NA  8.812780e-09           14
## 4             538  2.114490      NA 2.149518e-311           14
## 5             262  1.342387      NA  2.372477e-74           12
## 6             538  2.086431      NA 5.197609e-305           19
##   num_gene_query num_gene_hit  gene_hit log_ratio_gene pval_fisher_gene
## 1             77            2 3684;3383       4.320145     0.0003129003
## 2             77            2   721;720       4.233134     0.0003748918
## 3             77            2 3684;3383       4.078983     0.0005152773
## 4             77            2   721;720       4.078983     0.0005152773
## 5            108            2   721;720       3.894808     0.0007355142
## 6             77            2   721;720       3.773601     0.0009607091
\end{verbatim}

\begin{adjustwidth}{\fltoffset}{0mm}
\includegraphics{loci2path-vignette_files/figure-latex/unnamed-chunk-3-1} \end{adjustwidth}

\begin{adjustwidth}{\fltoffset}{0mm}
\includegraphics{loci2path-vignette_files/figure-latex/unnamed-chunk-3-2} \end{adjustwidth}

\begin{adjustwidth}{\fltoffset}{0mm}
\includegraphics{loci2path-vignette_files/figure-latex/unnamed-chunk-3-3} \end{adjustwidth}

\begin{verbatim}
## [1] "http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_MONOCYTE_PATHWAY"    
## [2] "http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_LECTIN_PATHWAY"      
## [3] "http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_GRANULOCYTES_PATHWAY"
## [4] "http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_CLASSIC_PATHWAY"     
## [5] "http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_LECTIN_PATHWAY"      
## [6] "http://www.broadinstitute.org/gsea/msigdb/cards/BIOCARTA_COMP_PATHWAY"
\end{verbatim}

\end{document}
