---
title: "Loci2path: regulatory annotation of genomic intervals based on tissue-specific expression QTLs"
author: 
  - Tianlei Xu
  - Zhaohui Qin
date: "`r doc_date()`"
package: "`r pkg_ver('loci2path')`"
abstract: > 
  Annotating a given genomic locus or a set of genomic loci is an important yet challenging task. This is especially true for the non-coding part of the genome which is enormous yet poorly understood. Since gene set enrichment analyses have demonstrated to be effective approach to annotate a set of genes, this idea can be extended to explore the enrichment of functional elements or features in a set of genomic intervals to reveal potential functional connections. In this study, we describe a novel computational strategy that takes advantage of the newly emerged, genome-wide and tissue-specific expression quantitative trait loci (eQTL) information to help annotate a set of genomic intervals in terms of transcription regulation. By checking the presence or absence of millions of eQTLs in the set of genomic intervals of interest, loci2path build a bridge connecting genomic intervals to biological pathway or pre-defined biological-meaningful gene sets. Our method enjoys two key advantages over existing methods: first, we no longer rely on proximity to link a locus to a gene which has shown to be unreliable; second, eQTL allows us to provide the regulatory annotation under the context of specific tissue types which is important.

output: 
  BiocStyle::pdf_document
  
vignette: >
  %\VignetteIndexEntry{loci2path}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  

bibliography: loci2path-vignette.bib
---



# Prepare input dataset for query

## Query regions

`loci2path` takes query regions in the format of `GenomicRanges`. Only the Genomic Locations (chromosomes, start and end position) will be used. Strand information and other metadata columns are ignored.
In the demo data, 47 regions associated with Psoriasis disease were downloaded from **immunoBase.org** and used as demo query regions. 

```{r query_region}
bed.file=system.file("extdata", "query/Psoriasis.BED", package = "loci2path")
query.bed=read.table(bed.file, header=F)
colnames(query.bed)=c("chr","start","end")
query.gr=makeGRangesFromDataFrame(query.bed)
```

## Prepare eQTL sets.

eQTL sets are entities recording 1-to-1 links between eQTL SNPs and genes. eQTL set entity also contains the following information: tissue name for the eQTL study, IDs and genomic ranges for the eQTL SNPs, IDs for the associated genes.

eQTL set can be constructed manually by specifying the corresponding information in each slot. 

eQTL set list is a list of multiple eQTL sets, usually collected from different tissues.

Below is an example to construct customized eQTL set and eQTL set list using demo data files. In the demo data folder, three eQTL sets downloaded from GTEx project are included. Due to the large size, each eQTL dataset is down sampled to 3000 records for demostration purpose. 

### construct eQTL set
```{r eset}
brain.file=system.file("extdata", "eqtl/brain.gtex.txt", package = "loci2path")
tab=read.table(brain.file, stringsAsFactors = F, header = T)
snp.gr=GRanges(seqnames=Rle(tab$snp.chr), 
  ranges=IRanges(start=tab$snp.pos, 
  width=1))
brain.eset=eqtlSet(tissue="brain",
  snp.id=tab$snp.id,
  snp.gr=snp.gr,
  gene=as.character(tab$gene.entrez.id))
brain.eset

skin.file=system.file("extdata", "eqtl/skin.gtex.txt", package = "loci2path")
tab=read.table(skin.file, stringsAsFactors = F, header = T)
snp.gr=GRanges(seqnames=Rle(tab$snp.chr), 
  ranges=IRanges(start=tab$snp.pos, 
  width=1))
skin.eset=eqtlSet(tissue="skin",
  snp.id=tab$snp.id,
  snp.gr=snp.gr,
  gene=as.character(tab$gene.entrez.id))
skin.eset

blood.file=system.file("extdata", "eqtl/blood.gtex.txt", package = "loci2path")
tab=read.table(blood.file, stringsAsFactors = F, header = T)
snp.gr=GRanges(seqnames=Rle(tab$snp.chr), 
  ranges=IRanges(start=tab$snp.pos, 
  width=1))
blood.eset=eqtlSet(tissue="blood",
  snp.id=tab$snp.id,
  snp.gr=snp.gr,
  gene=as.character(tab$gene.entrez.id))
blood.eset
```

### construct eQTL set list

```{r esetlist}
eset.list=list(Brain=brain.eset, Skin=skin.eset, Blood=blood.eset)
eset.list
```



## Prepare gene set collection

A geneset collection contains a list of gene sets, with each gene set is represented as a vector of member genes. A vector of description is also provided as the metadata slot for each gene set. The total number of gene in the geneset collection is also required to perform the enrichment test. In this tutorial the BIOCARTA pathway collection was downloaded from MSigDB. 

```{r}
biocarta.link.file=system.file("extdata", "geneSet/biocarta.txt", package = "loci2path")
biocarta.set.file=system.file("extdata", "geneSet/biocarta.set.txt", package = "loci2path")

biocarta.link=read.delim(biocarta.link.file, header = F, stringsAsFactors = F)
set.geneid=read.table(biocarta.set.file, stringsAsFactors = F)
set.geneid=strsplit(set.geneid[,1], split=",")
names(set.geneid)=biocarta.link[,1]

head(biocarta.link)
head(set.geneid)
```

In order to build gene set, we also need to know the total number of genes in order to perform enrichment test. In this study, the total number of gene in MSigDB pathway collection is 31,847[@Liberzon2015]
```{r}
#build geneSet
biocarta=geneSet(
  gene.set=set.geneid,
  description=biocarta.link[,2],
  total.number.gene=31847)
biocarta
```






# Perform query

# explore query result

## extract tissue-pathway heatmap

## extract word cloud from result

## obtain eQTL gene list

## obtain average tissue degree for each pathway

## obtain tissue enrichment for query regions





```{r echo = FALSE}

#show query region
query.gr
#show eqtl set
eqtl.set.list
#show gene set(biocarta)
biocarta

#query
result=query.egset.list(query.gr=query.gr, query.score=NULL, eqtl.set.list=eqtl.set.list, gene.set=biocarta)
head(result)
result.parallel=query.egset.list(query.gr=query.gr, query.score=NULL, eqtl.set=eqtl.set.list, gene.set=biocarta,
                                 parallel = T)
head(result.parallel)


#query tissue specificity
gr.tissue=query.tissue(query.gr, eqtl.set.list=eqtl.set.list)

#reorder result by tissue specificity
result2=res.order.by.tissue(result, tissue.list = rownames(gr.tissue))

#extract tissue/geneset matrix
mat=res.get.heat.mat(result, test.method = "fisher")

#plot heatmap
draw.heatmap(mat)

#plot word cloud
draw.wordcloud(result)

#plot p-value distribution of result
draw.pval.distribution(result, test.method="fisher")

#obtain geneset description from object
description=get.geneset.description(biocarta, geneset.ids=result$name_pthw)
head(description)
```


# References

